python/ray/data/_internal/util.py

```diff
@@ -502,37 +504,105 @@ def _split_list(arr: List[Any], num_splits: int) -> List[List[Any]]:
     return splits
 
 
-def validate_compute(
+def get_compute_strategy(
     fn: "UserDefinedFunction",
-    compute: Optional[Union[str, "ComputeStrategy"]],
     fn_constructor_args: Optional[Iterable[Any]] = None,
-) -> None:
+    compute: Optional[Union[str, "ComputeStrategy"]] = None,
+    concurrency: Optional[Union[int, Tuple[int, int]]] = None,
+) -> "ComputeStrategy":
+    """Get `ComputeStrategy` based on the function or class, and concurrency
+    information.
+
+    Args:
+        fn: The function or generator to apply to a record batch, or a class type
+            that can be instantiated to create such a callable.
+        fn_constructor_args: Positional arguments to pass to ``fn``'s constructor.
+        compute: Either "tasks" (default) to use Ray Tasks or an
+                :class:`~ray.data.ActorPoolStrategy` to use an autoscaling actor pool.
+        concurrency: The number of Ray workers to use concurrently.
+
+    Returns:
+       The `ComputeStrategy` for execution.
+    """
     # Lazily import these objects to avoid circular imports.
     from ray.data._internal.compute import ActorPoolStrategy, TaskPoolStrategy
-    from ray.data.block import CallableClass
 
-    if isinstance(fn, CallableClass) and (
-        compute is None or compute == "tasks" or isinstance(compute, TaskPoolStrategy)
-    ):
-        raise ValueError(
-            "``compute`` must be specified when using a CallableClass, and must "
-            f"specify the actor compute strategy, but got: {compute}. "
-            "For example, use ``compute=ray.data.ActorPoolStrategy(size=n)``."
+    # Check if `fn` is a function or not.
+    # NOTE: use `inspect.isfunction(fn)` instead of `instanceof(fn, CallableClass)`,
+    # because latter returns False for an object instance of callable class.
+    if inspect.isfunction(fn):
```

Comment:
This is the major behavior change. Previously we rely on checking if `isinstance(fn, CallableClass)`, but it's not right when creating an object for the callable class. Changed to use `inspect.isfunction(fn)` to check if it's a function, which is more stable. Added a unit test to cover it.

```py
import ray
from ray.data.block import CallableClass

def foo(x):
    return x

class BarClass:
    def __init__(self, x):
        self._x = x
    def __call__(self, x):
        return x

>>> isinstance(foo, CallableClass)
False
>>> isinstance(BarClass, CallableClass)
True
>>> isinstance(BarClass(1), CallableClass)
False # <-- This does not work with current logic
```

Reply:
@ericl - ah that makes sense. This was coming from one of test failure - https://github.com/ray-project/ray/blob/master/doc/source/ray-core/_examples/datasets_train/datasets_train.py#L702 . I changed the test code instead.

btw on master, we always allow to run this kind of object instance (as a function):

```py
>>> class BarClass:
...     def __init__(self, x):
...         self._x = x
...     def __call__(self, x):
...         return x
... 
>>> 
>>> fn = BarClass(1)
>>> ds = ray.data.range(10)
>>> ds = ds.map(fn)
>>> ds.take_all()
[{'id': 0}, {'id': 2}, {'id': 3}, {'id': 5}, {'id': 7}, {'id': 8}, {'id': 9}, {'id': 1}, {'id': 4}, {'id': 6}]
```

Are you thinking we should explicitly disallow and throw error if user passes an object instance? I am a little hesitating given this is always allowed.

Reply:
I see, in this case, I think we should probably raise an error since it's definitely not doing what the user is expecting--- the `__init__` won't be cached even though it looks like they're passing in an actor class.

Reply:
For the Preprocessor classes, you could wrap the class method in a lambda to fix?

Reply:
for our own code, we can work around with lambda, given we know the function signature, and what arguments to pass. What about user-defined arbitrary class method?